internal/buffer/settings.go

```diff
@@ -8,6 +8,43 @@ import (
 	"github.com/zyedidia/micro/v2/internal/screen"
 )
 
+func (b *Buffer) ReloadSettings(filetype bool) {
+	settings := config.ParsedSettings()
+
+	if _, ok := b.LocalSettings["filetype"]; ok || filetype {
+		settings["filetype"] = b.Settings["filetype"]
+	} else if _, ok := settings["filetype"]; ok && settings["filetype"] != "unknown" {
+		b.Settings["filetype"] = settings["filetype"]
+	} else {
+		b.Settings["filetype"] = "unknown"
+	}
+	if b.Settings["filetype"] == "unknown" {
+		b.UpdateRules()
```

Comment:
We also update rules (unconditionally) in `DoSetOptionNative("filetype")` and in `reload` (and we indeed should do that, especially in `reload`, since it may reload syntax files). So we unnecessarily update them twice?

So maybe we should instead update rules just here, in `ReloadSettings()`, unconditionally, but before updating `settings["filetype"]`?

i.e. something like:

```
	if _, ok := b.LocalSettings["filetype"]; !ok && !filetype {
		// need to update filetype from parsed settings
		b.Settings["filetype"] = "unknown"
		if v, ok := settings["filetype"]; ok {
			b.Settings["filetype"] = v
		}
	}

	// update syntax rules, which will also update filetype if needed
	b.UpdateRules()
	settings["filetype"] = b.Settings["filetype"]
```

Reply:
Also maybe rename the `filename` argument to `reloadFilename` (and inverse its value: `reloadRuntime()` would call `ReloadSettings()` with reloadFilename=true, while `DoSetOptionNative("filetype")` would call it with reloadFilename=false)?

Reply:
Very good idea to simplify the logic, besides the small typo `!filetype` -> `filetype` in `if _, ok := b.LocalSettings["filetype"]; !ok && !filetype {`. :wink:

Reply:
FWIW it wasn't a typo.